Call IDataWriteProcessor on all data element mutations and add new POST endpoint with better response type

ivarne opened the pull request:
This should be simple and (mostly)backwards compatible changes before adding new hooks for data processing and file validation.

The commits are fairly well separated, so they might be reviewed individually if you want smaller chunks.


## Related Issue(s)
- #850 

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

ivarne committed:
Remove unused property on DataPatchResult

ivarne committed:
Rename CachedInstanceDataAccessor to InstanceDataUnitOfWork

Also take ApplicationMetadata instead of IAppMeatada service in constructor (it is usually already availible)

ivarne committed:
Don't throw when a page can't be evaluated because no data elements of the default type exists.

ivarne committed:
Add HttpContextExtension method to read request as byte[] efficiently

ivarne committed:
Cleanup

* Better errro handling in ModelSerializationService
* make DataModelPairResponse a separate class, instead of a nested class
* Add JsonPropertyName on ValidationSourcePair

ivarne committed:
Refactor DataController.Post

* Split into two endpoints (for compatibility)
* Old endpoint with dataType as GET param is unchanged.
* New endpoint returns all app state, the same way PATCH does
* Extend ProblemDetails explicitly so that swagger is correct

martinothamar commented on src/Altinn.App.Core/Internal/Patch/PatchService.cs:
```diff
@@ -202,6 +203,24 @@ await RunDataProcessors(
         };
     }
 
+    public async Task<List<ValidationSourcePair>> RunIncrementalValidation(
```
What's the rationale for adding this?

ivarne replied:
Just that I did not want to inject the full ValidationService twice for PatchService and directly in the controller.

martinothamar replied:
Would prefer to just inject another service tbh 😄

ivarne replied:
Yes, when we get your patch to dynamically initialise the validators, so that it don't happen twice for every request (service owners does strange things in their constructors), it would be my preference too.

Actually I would prefer to not have PatchService in the first place, because it doesn't have natural boundaries, but it was added in code review to make DataController shorter 🥲.

HauklandJ reviewed (approved):
:raised_hands:

ivarne committed:
Ensure that we cleanup instances after running tests that create instances

ivarne committed:
Rewrite interfaces IDataWriteProcessor and IValidator (Bulk Changes test:fail)

List<DataElementChange> gets wrapped in DataElementChanges to support class hierarchy of Change types

Don't send Instance in addition to IInstanceDataAccessor

ivarne committed:
Use IDataWriteProcessor in all apis to mutate data elements.

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/better-file-upload-validation

ivarne committed:
Fix sonar issues

martinothamar commented on src/Altinn.App.Core/Models/DataElementChanges.cs:
```diff
@@ -0,0 +1,129 @@
+/// <summary>
+/// Representation of a change to a binary data element
+/// </summary>
+public sealed class BinaryChange : DataElementChange
```
`BinaryDataChange` ?

ivarne replied:
🎉 👍

ivarne committed:
Final clanup for v8.5

ivarne committed:
Rename BinaryChange => BinaryDataChange
